risc0/zkvm/src/exec/mod.rs

```diff
@@ -159,6 +159,15 @@ impl<'a> Executor<'a> {
     /// Run the executor until [ExitCode::Paused] or [ExitCode::Halted] is
     /// reached, producing a [Session] as a result.
     pub fn run(&mut self) -> Result<Session> {
+        self.run_with_callback(|segment| Box::new(SimpleSegmentRef::new(segment)))
+    }
+
+    /// Run the executor until [ExitCode::Paused] or [ExitCode::Halted] is
+    /// reached, producing a [Session] as a result.
+    pub fn run_with_callback<F>(&mut self, mut callback: F) -> Result<Session>
+    where
+        F: FnMut(Segment) -> Box<dyn SegmentRef>,
```

Comment:
Should like be a result? I'm guessing that that a callback may, for instance, upload the `Segment` to an S3 bucket, or otherwise store it at some network accessible location.
```suggestion
        F: FnMut(Segment) -> Result<Box<dyn SegmentRef>>,
```

Reply:
Another option on this would be to make either `Executor` or `Executor::run` generic over some kind of `SegmentStorage` trait. It would functionally be the same, just using an object instead of a closure. E.g. 

```rust
trait SegmentStore {
    type Ref: SegmentRef;

    fn store(&mut self, segment: Segement) -> Result<Self::Ref>;
}

pub struct SimpleSegmentStore;

impl SegmentRef for Segment {
    fn resolve(&self) -> Result<Segment> { Ok(self.clone()) }
}

impl SegmentStorage for SimpleSegmentStore {
    type Ref = Segment;

    fn store(&mut self, segment: Segement) -> Result<Self::Ref> { Ok(Segment) };
}

impl Executor {
...
    pub fn run(&mut self) -> Result<Session> {
        self.run_with_store(SimpleSegmentStore)
    }

    pub fn run_with_callback(&mut self, callback: &mut impl SegmentStore) -> Result<Session> {
        ...
    }
...
}
```

Reply:
Yes, for sure the callback should return a `Result`. `SegmentStorage` is an interesting idea.

Reply:
After thinking about this, I like the callback a little better because it allows more choice by the caller. If the caller wants to use a trait, they can do that and wire up the callback to a function on the trait. But from the Executor's point of view, only a call is needed. I would go with this trait being part of the `run_with_callback` if there was more than just a call that was needed.